File: scripts_opcionais/classify_inbox.py
import os
import json
import re
from datetime import datetime

# Caminhos
BASE_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
INBOX_DIR = os.path.join(BASE_DIR, "05_Biblioteca_IA/inbox")
TAX_PATH = os.path.join(BASE_DIR, "05_Biblioteca_IA/data/biblioteca_taxonomia_temas.json")
CAT_OUT = os.path.join(BASE_DIR, "05_Biblioteca_IA/data/biblioteca_catalogo.json")
BRAIN_OUT = os.path.join(BASE_DIR, "05_Biblioteca_IA/data/biblioteca_brain_connections.json")

# Carregar taxonomia
with open(TAX_PATH, 'r', encoding='utf-8') as f:
    taxonomy = json.load(f)

themes = taxonomy['themes']

# Listar arquivos do inbox
files = [f for f in os.listdir(INBOX_DIR) if f not in ['.DS_Store', 'README.md'] and os.path.isfile(os.path.join(INBOX_DIR, f))]

print(f"📂 Total de arquivos no inbox: {len(files)}")

def get_type(name):
    ext = os.path.splitext(name)[1].lower().replace('.', '')
    if ext in ['pdf']: return 'pdf'
    if ext in ['docx', 'doc']: return 'word'
    if ext in ['html', 'htm']: return 'html'
    if ext in ['md', 'txt']: return 'text'
    if ext in ['csv', 'xlsx']: return 'data'
    if ext in ['png', 'jpg', 'jpeg', 'webp']: return 'image'
    return 'outro'

def classify(filename):
    clean_name = re.sub(r'[_.-]', ' ', filename).lower()
    best_theme = None
    best_score = 0
    
    for theme in themes:
        if theme['id'] == 'inbox-revisar': continue
        score = 0
        for kw in theme['keywords']:
            if kw.lower() in clean_name:
                score += len(kw)
        
        if score > best_score:
            best_score = score
            best_theme = theme
            
    # Score de corte: pelo menos 3 caracteres batendo
    if best_score >= 3:
        return best_theme
    return next(t for t in themes if t['id'] == 'inbox-revisar')

def to_slug(name):
    name_no_ext = os.path.splitext(name)[0]
    slug = re.sub(r'[^a-zA-Z0-9]', '-', name_no_ext).lower()
    slug = re.sub(r'-+', '-', slug).strip('-')
    return slug[:80]

def to_title(name):
    name_no_ext = os.path.splitext(name)[0]
    title = re.sub(r'[_.-]+', ' ', name_no_ext)
    return title.title().strip()

items = []
nodes = [
    { "id": "biblioteca-ia", "label": "📚 Biblioteca IA", "type": "module", "theme": "projeto", "path": "05_Biblioteca_IA/index.html" }
]
edges = []
stats = {}

for file in files:
    theme = classify(file)
    tipo = get_type(file)
    slug = to_slug(file)
    title = to_title(file)
    theme_id = theme['id']
    
    stats[theme_id] = stats.get(theme_id, 0) + 1
    
    items.append({
        "id": slug,
        "title": title,
        "filename": file,
        "theme": theme_id,
        "tipo": tipo,
        "tags": [theme_id, tipo],
        "status": "revisar" if theme_id == 'inbox-revisar' else "catalogado",
        "path": "inbox/" + file,
        "resumo": "",
        "ia_origem": "Auto-classificado",
        "data": "2026-05",
        "createdAt": datetime.now().isoformat()
    })
    
    # Brain node
    nodes.append({
        "id": slug,
        "label": (theme['emoji'] + ' ' if theme else '📥 ') + title[:50],
        "type": "file",
        "theme": theme_id,
        "path": "inbox/" + file
    })
    
    # Brain edges
    edges.append({ "from": slug, "to": "biblioteca-ia", "relation": "pertence-a" })
    edges.append({ "from": slug, "to": theme_id, "relation": "classificado-em" })
    
    # Extra connections
    if theme_id == 'temi-prova' or 'temi' in file.lower():
        edges.append({ "from": slug, "to": "banco-temi", "relation": "gera-questoes" })
    if theme_id == 'farmaco-doses' or any(kw in file.lower() for kw in ['dose', 'calculo', 'dilu']):
        edges.append({ "from": slug, "to": "calculadoras-uti", "relation": "apoia-calculo" })

# Add theme nodes
for theme in themes:
    nodes.append({
        "id": theme['id'],
        "label": theme['emoji'] + ' ' + theme['label'],
        "type": "theme",
        "theme": theme['id'],
        "path": theme['folder']
    })

# Save catalog
catalog = {
    "updatedAt": datetime.now().strftime("%Y-%m-%d"),
    "project": "Biblioteca IA — Enciclopédia Médica Intensiva",
    "totalFiles": len(items),
    "stats": stats,
    "items": items
}
with open(CAT_OUT, 'w', encoding='utf-8') as f:
    json.dump(catalog, f, indent=2, ensure_ascii=False)

# Save brain
brain = {
    "updatedAt": datetime.now().strftime("%Y-%m-%d"),
    "nodes": nodes,
    "edges": edges
}
with open(BRAIN_OUT, 'w', encoding='utf-8') as f:
    json.dump(brain, f, indent=2, ensure_ascii=False)

print(f"\n✅ Catálogo gerado: {len(items)} arquivos")
print(f"📊 Classificação por tema:")
sorted_stats = sorted(stats.items(), key=lambda x: x[1], reverse=True)
for theme_id, count in sorted_stats:
    t = next((th for th in themes if th['id'] == theme_id), None)
    emoji = t['emoji'] if t else '📥'
    print(f"   {emoji} {theme_id}: {count}")

print(f"\n🧠 Brain: {len(nodes)} nós, {len(edges)} edges")
print(f"📄 Salvos em:")
print(f"   {CAT_OUT}")
print(f"   {BRAIN_OUT}")

File: scripts_admin/scripts_opcionais/classify_inbox.js
// Script de auto-classificação dos arquivos do inbox
// Lê os arquivos, compara com keywords da taxonomia, gera catalogo + brain connections
const fs = require('fs');
const path = require('path');

const INBOX = path.join(__dirname, '05_Biblioteca_IA/inbox');
const TAX_PATH = path.join(__dirname, '05_Biblioteca_IA/data/biblioteca_taxonomia_temas.json');
const CAT_OUT = path.join(__dirname, '05_Biblioteca_IA/data/biblioteca_catalogo.json');
const BRAIN_OUT = path.join(__dirname, '05_Biblioteca_IA/data/biblioteca_brain_connections.json');

const taxonomy = JSON.parse(fs.readFileSync(TAX_PATH, 'utf8'));
const themes = taxonomy.themes;

// Ler arquivos do inbox
const files = fs.readdirSync(INBOX).filter(f => {
  if (f === '.DS_Store' || f === 'README.md') return false;
  return true;
});

console.log(`📂 Total de arquivos no inbox: ${files.length}`);

// Determinar tipo
function getType(name) {
  const ext = path.extname(name).toLowerCase().replace('.', '');
  if (['pdf'].includes(ext)) return 'pdf';
  if (['docx', 'doc'].includes(ext)) return 'word';
  if (['html', 'htm'].includes(ext)) return 'html';
  if (['md', 'txt'].includes(ext)) return 'text';
  if (['csv', 'xlsx'].includes(ext)) return 'data';
  if (['png', 'jpg', 'jpeg', 'webp'].includes(ext)) return 'image';
  return 'outro';
}

// Classificar por keywords
function classify(filename) {
  const lower = filename.toLowerCase().replace(/[_\-\.]/g, ' ');
  let bestTheme = null;
  let bestScore = 0;
  
  for (const theme of themes) {
    if (theme.id === 'inbox-revisar') continue;
    let score = 0;
    for (const kw of theme.keywords) {
      if (lower.includes(kw.toLowerCase())) {
        score += kw.length; // Longer keyword matches score more
      }
    }
    if (score > bestScore) {
      bestScore = score;
      bestTheme = theme;
    }
  }
  
  return bestScore >= 3 ? bestTheme : themes.find(t => t.id === 'inbox-revisar');
}

// Gerar slug
function toSlug(name) {
  return name.replace(/\.[^.]+$/, '')
    .replace(/[^a-zA-Z0-9]/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '')
    .toLowerCase()
    .slice(0, 80);
}

// Gerar título
function toTitle(name) {
  return name.replace(/\.[^.]+$/, '')
    .replace(/[_\-]+/g, ' ')
    .replace(/\b\w/g, c => c.toUpperCase())
    .trim();
}

// Processar
const items = [];
const nodes = [
  { id: 'biblioteca-ia', label: '📚 Biblioteca IA', type: 'module', theme: 'projeto', path: '05_Biblioteca_IA/index.html' }
];
const edges = [];
const stats = {};

for (const file of files) {
  const theme = classify(file);
  const tipo = getType(file);
  const slug = toSlug(file);
  const title = toTitle(file);
  const themeId = theme ? theme.id : 'inbox-revisar';
  
  stats[themeId] = (stats[themeId] || 0) + 1;
  
  items.push({
    id: slug,
    title: title,
    filename: file,
    theme: themeId,
    tipo: tipo,
    tags: [themeId, tipo],
    status: themeId === 'inbox-revisar' ? 'revisar' : 'catalogado',
    path: 'inbox/' + file,
    resumo: '',
    ia_origem: 'Auto-classificado',
    data: '2026-05',
    createdAt: new Date().toISOString()
  });
  
  // Brain node
  nodes.push({
    id: slug,
    label: (theme ? theme.emoji + ' ' : '📥 ') + title.slice(0, 50),
    type: 'file',
    theme: themeId,
    path: 'inbox/' + file
  });
  
  // Brain edges
  edges.push({ from: slug, to: 'biblioteca-ia', relation: 'pertence-a' });
  edges.push({ from: slug, to: themeId, relation: 'classificado-em' });
  
  // Extra connections
  if (['temi-prova'].includes(themeId) || file.toLowerCase().includes('temi')) {
    edges.push({ from: slug, to: 'banco-temi', relation: 'gera-questoes' });
  }
  if (['farmaco-doses'].includes(themeId) || file.toLowerCase().includes('dose') || file.toLowerCase().includes('calculo')) {
    edges.push({ from: slug, to: 'calculadoras-uti', relation: 'apoia-calculo' });
  }
}

// Add theme nodes
for (const theme of themes) {
  nodes.push({
    id: theme.id,
    label: theme.emoji + ' ' + theme.label,
    type: 'theme',
    theme: theme.id,
    path: theme.folder
  });
}

// Save catalog
const catalog = {
  updatedAt: new Date().toISOString().slice(0, 10),
  project: 'Biblioteca IA — Enciclopédia Médica Intensiva',
  totalFiles: items.length,
  stats: stats,
  items: items
};
fs.writeFileSync(CAT_OUT, JSON.stringify(catalog, null, 2), 'utf8');

// Save brain
const brain = {
  updatedAt: new Date().toISOString().slice(0, 10),
  nodes: nodes,
  edges: edges
};
fs.writeFileSync(BRAIN_OUT, JSON.stringify(brain, null, 2), 'utf8');

// Report
console.log(`\n✅ Catálogo gerado: ${items.length} arquivos`);
console.log(`📊 Classificação por tema:`);
const sorted = Object.entries(stats).sort((a, b) => b[1] - a[1]);
for (const [theme, count] of sorted) {
  const t = themes.find(th => th.id === theme);
  console.log(`   ${t ? t.emoji : '📥'} ${theme}: ${count}`);
}
console.log(`\n🧠 Brain: ${nodes.length} nós, ${edges.length} edges`);
console.log(`📄 Salvos em:`);
console.log(`   ${CAT_OUT}`);
console.log(`   ${BRAIN_OUT}`);
